Sort clients and servers on admin dashboard

klingerf opened the pull request:
Sorts clients and servers on the admin dashboard alphabetically by their label. The values are pulled from a map, and without sorting they show up in arbitrary order. After the page loads, as new clients are added they are also inserted alphabetically into the page, and the coloring of the existing clients is redrawn accordingly.

As part of this branch, I'm also sorting routers in the order that they're defined in the config file, and I'm removing the old metrics handler and all of its assets, since that page is no longer accessible in the admin dashboard.

Fixes #915.

rmars commented:
Looks awesome, thanks for doing this.

I think we'll also need to add something in `router_clients.js` - when new clients show up they'll just be appended to the dom... then the clients won't appear in sorted order.  

So we'd need to clear and reappend all the `client-container`s. This could be expensive if there's a lot of clearing and reappending going on so maybe that's not worth it, since you could always refresh to get a sorted page. (Though maybe we don't want people to have to do this).


example:
add a bunch of clients
![screen shot 2017-01-11 at 5 13 45 pm](https://cloud.githubusercontent.com/assets/549258/21873192/c8ae3a30-d821-11e6-85ca-3e73eb360ec4.png)

need to refresh to resort:
![screen shot 2017-01-11 at 5 14 39 pm](https://cloud.githubusercontent.com/assets/549258/21873195/cfae0bbc-d821-11e6-92b3-262ba5b4b385.png)

rmars commented:
I think we'll also need to change `assignColorsToClients` in `router-clients.js`(my fault for making it order-dependent) - the reordering is causing color mapping issues, were more than 1 client is assigned the same color, even though we have more colors available:

![screen shot 2017-01-11 at 5 21 04 pm](https://cloud.githubusercontent.com/assets/549258/21873363/ec1fe7ce-d822-11e6-9af0-670f08144747.png)

adleong commented:
@rmars is is possible to insert the new client into the correct position in the dom instead of clearing and re-appending?

rmars commented:
yeah! good point! we can use $.insertBefore and the like to just do inserts

klingerf commented:
@rmars Thanks for the suggestion! I added client sort-on-update in 7504903c. I also removed a few things that appeared to be unused, but please have a look to make sure I didn't break anything.

adleong commented:
Does this mean that when new clients are added, existing clients might change color?

klingerf commented:
@adleong That's correct.

adleong commented on admin/src/main/resources/io/buoyant/admin/js/src/logger-overrides.js:
```diff
@@ -12,6 +12,7 @@ $(function() {
   '        <div id="navbar" class="collapse navbar-collapse">' +
   '          <ul class="nav navbar-nav">' +
   '            <li><a href="../delegator">dtab</a></li>' +
+  '            <li><a href="../requests">requests</a></li>' +
```
ahh, good catch.  thanks for fixing!

siggy replied:
consider a comment here and/or in `AdminHandler.scala` that the nav's should stay in sync.

rmars replied:
(I'm working on that right now as part of #505!)

siggy commented on admin/src/main/resources/io/buoyant/admin/js/src/router_clients.js:
```diff
@@ -13,12 +13,13 @@ define([
-        clientMapping[client.label] = colors[colorIdx++ % colors.length];
-        return clientMapping;
-      }, {});
+      _(clients).sortBy('label').forEach(function(client, idx) {
```
there's a `return _.sortBy(clients, 'label');` on line 204 of `routers.js`. is this sorting twice?

siggy commented on admin/src/main/resources/io/buoyant/admin/js/src/router_clients.js:
```diff
@@ -70,15 +79,15 @@ define([
         });
 
-        var expandClients = shouldExpandClients(addedClients.length + routerClients.length);
+        expandClients = shouldExpandClients(routers.clients(routerName).length);
```
nit: i see `routers.clients(routerName)` repeated (and also assignned to `clients`), consider DRY up.

klingerf commented:
@adleong Ok, per discussion, have updated the behavior to append new clients and preserve the color of existing clients.

adleong commented:
👏 👏 👏  thank you, @klingerf!  Sorry for the runaround
